### PR TITLE
[Chore] Add weeder check

### PR DESCRIPTION
**Pull request by Sereja313**
Problem: We want to be able to check our Haskell applications for dead code. This can be done using weeder.

Solution: Add weeder check to our overlay.

### CODE REVIEW

File: overlay/default.nix [hunk 1]
```diff
@@ -124,6 +124,36 @@ in
         cd ${src}
         ${final.cabal-install}/bin/cabal check
       '';
+
+      weeder = { weederToml, hs-pkgs }: let
```
Comment (rvem) [inline overlay/default.nix:128]: I propose to leave a comment that `writeHieFiles = true` is mandatory for packages that are supposed to be checked by weeder

File: overlay/default.nix [hunk 2]
```diff
@@ -124,6 +124,36 @@ in
...
+        haskellPkgs = inputs.haskell-nix.legacyPackages.x86_64-linux;
+
+        # get local package names
+        pkg-names = lib.attrNames (lib.filterAttrs (_: v: (builtins.isAttrs v) && v ? isLocal && v.isLocal) hs-pkgs);
+
+        # get all hie files
+        allHieFiles = final.linkFarmFromDrvs "all-hie-files" (lib.concatMap (name: let
+          hs-pkg = hs-pkgs.${name}.components;
+        in (map (component: component.hie) (
```
Comment (rvem) [inline overlay/default.nix:137]: Will this fail with an error in case `writeHieFiles = true` wasn't provided?

File: flake.nix [hunk 3]
```diff
@@ -19,6 +19,10 @@
       url = "github:Kleidukos/get-tested/v0.1.6.0";
       flake = false;
     };
+    weeder-src = {
+      url = "github:ocharles/weeder";
+      flake = false;
```
Comment (rvem) [inline flake.nix:24]: In fact, it is actually a [flake](https://github.com/ocharles/weeder/blob/master/flake.nix) :pled:
Sadly, it doesn't expose executable

**Review by rvem — APPROVED** [review]

**Comment by gromakovsky** [comment]
@Sereja313 or @rvem could you please look into it? :arrow_up:

**Comment by Sereja313** [comment]
> Sorry for late review, but recently there appeared https://github.com/NorfairKing/weeder-nix which is now mentioned in https://github.com/ocharles/weeder README. Can we use it to simplify the code written in this PR?

The current solution is based on the `haskell.nix`, which we already use for building and testing package, and `weeder-nix` uses the `nixpkgs` toolkit to generate hie files. So I'm guessing there may be inconsistencies when the package builds with `haskell.nix`, but it also requires some manual work to make it suitable for building with `nixpkgs` to get `weeder-nix` to work.
So I'm not sure if it's worth it

**Review by gromakovsky — APPROVED** [review]

**Pull request merged**